nixos/nginx: allow arbitrary parameters in upstream servers

Izorkin opened the pull request:
###### Description of changes
Add extraOptions in upstream configuration

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -105,7 +105,7 @@ let
     upstream ${name} {
       ${toString (flip mapAttrsToList upstream.servers (name: server: ''
-        server ${name} ${optionalString server.backup "backup"};
+        server ${name} ${optionalString (server.failTimeout != "10") "fail_timeout=${server.failTimeout}"} ${optionalString server.backup "backup"};
```
Shouldn't we introduce a generic option to add potential other arguments?

Izorkin replied:
Would there be any confusion with `${upstream.extraConfig}`

Izorkin replied:
Addede `extraOptions`.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2143

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -105,7 +105,7 @@ let
     upstream ${name} {
       ${toString (flip mapAttrsToList upstream.servers (name: server: ''
-        server ${name} ${optionalString server.backup "backup"};
+        server ${name} ${optionalString server.backup "backup"} ${optionalString (server.extraOptions != []) "${concatStringsSep " " server.extraOptions}"};
```
Shouldn't we migrate server.back to this option?

Izorkin replied:
It can be done. I wanted to do PR with minimal changes, so as not to break existing configurations.

Izorkin replied:
I was also unable to rename the parameter:
```
    (mkRenamedOptionModule [ "services" "nginx" "upstreams" "" "servers" "backup" ] [ "services" "nginx" "upstreams" "" "servers" "extraOptions" ])
```
```
error: The option `services.nginx.upstreams' in module `/home/user/nixpkgs/nixos/modules/services/web-servers/nginx/default.nix' would be a parent of the following options, but its type `attribute set of (submodule)' does not support nested options.
        - option(s) with prefix `services.nginx.upstreams.""' in module `/home/lafiel/nixpkgs/nixos/modules/services/web-servers/nginx/default.nix'
(use '--show-trace' to show detailed location information)
building Nix...
```

ncfavier commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -923,13 +923,11 @@ in
             servers = mkOption {
               type = types.attrsOf (types.submodule {
                 options = {
-                  backup = mkOption {
-                    type = types.bool;
-                    default = false;
-                    description = lib.mdDoc ''
-                      Marks the server as a backup server. It will be passed
-                      requests when the primary servers are unavailable.
-                    '';
```
We should keep `backup` and add a `freeformType = attrOf (oneOf [ bool int str ]);` for the other options. Refer to the [documentation](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#server) for what kinds of parameters the `server` directive accepts.

Izorkin replied:
Returned `backup` option.
Error with use `freeformType = with types; attrOf (oneOf [ bool int str ]);`:
```
       error: undefined variable 'attrOf'

       at /nix/store/qx77h4vzv2j8lp9s1mvm1xlda8cq45zh-nixos/modules/services/web-servers/nginx/default.nix:925:44:

          924|               type = types.attrsOf (types.submodule {
          925|                 freeformType = with types; attrOf (oneOf [ bool int str ]);
             |                                            ^
          926|                 options = {

```

ncfavier replied:
`attrsOf`

Izorkin replied:
Thanks, fixed!

ncfavier replied:
Sorry, I wasn't clear: I'm suggesting that we get rid of `extraOptions` and instead use the submodule itself to set `server` parameters, using a function like this one:

```nix
let toParameter = key: value:
  if builtins.isBool value
  then lib.optionalString value key
  else "${key}=${toString value}";
```

Then use it to render the `server` directive as follows:

```nix
''
  server ${name} ${concatStringsSep " " (mapAttrsToList toParameter server)};
''
```

The examples would then look something like this:

```nix
{
  servers."backend1.example.com:8080" = { weight = 5; };
  servers."backend2.example.com" = { max_fails = 3; fail_timeout = "30s"; };
  servers."backend3.example.com" = {};
}
```

Izorkin replied:
Reworked PR.

Izorkin committed:
nixos/nginx: allow arbitrary parameters in upstream servers

Adds a `freeformType` to `services.nginx.upstreams.<name>.servers` to
allow setting arbitrary parameters.

ncfavier reviewed (approved):
I've re-added `backup` and made a documentation pass. Could you test this please?

Izorkin commented:
> I've re-added `backup` and made a documentation pass.

Without a separate `backup` variant `servers = { "backup.example.com" = { backup = true; }; };` works. Does it make sense to keep it?

ncfavier commented:
The option adds typing information and documentation. It's not essential but, since it's already there, there's no point in removing it.

Izorkin commented:
> Could you test this please?

Thanks, working.
